src/lib/modes/aead/ascon_aead128/ascon_aead128.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+* Ascon-AEAD128 AEAD
+* (C) 2025 Jack Lloyd
+*     2025 René Meusel
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/ascon_aead128.h>
+
+#include <botan/internal/loadstor.h>
+#include <botan/internal/stl_util.h>
+
+namespace Botan {
+
+namespace {
+
+constexpr void xor2x64(std::span<uint64_t, 2> lhs, std::span<const uint64_t, 2> rhs) {
+   lhs[0] ^= rhs[0];
+   lhs[1] ^= rhs[1];
+}
+
+template <size_t N>
+constexpr auto as_array_of_uint64(std::span<const uint8_t> in) {
+   BOTAN_DEBUG_ASSERT(in.size() == N * sizeof(uint64_t));
+   return load_le<std::array<uint64_t, N>>(in.first<N * 8>());
+}
+
+// NIST SP.800-232 Appendix B (Table 13)
+constexpr Ascon_p initial_state_of_ascon_aead_permutation({
+   .init_and_final_rounds = 12,
+   .processing_rounds = 8,
+   .bit_rate = 128,
+   .initial_state = {},
+});
+
+// NIST SP.800-232 Section 5.1
+constexpr uint64_t ascon_aead_128_iv = 0x00001000808c0001;
+
+// NIST SP.800-232 Appendix A.2
+constexpr uint64_t ascon_aead_128_domain_sep = 0x8000000000000000;
+
+}  // namespace
+
+Ascon_AEAD128_Mode::Ascon_AEAD128_Mode() : m_ascon_p(initial_state_of_ascon_aead_permutation) {}
+
+void Ascon_AEAD128_Mode::clear() {
+   m_key.reset();
+   m_ad.clear();
+   reset();
+}
+
+void Ascon_AEAD128_Mode::reset() {
+   m_ascon_p = initial_state_of_ascon_aead_permutation;
+   m_started = false;
+   m_has_nonce = false;
+}
+
+void Ascon_AEAD128_Mode::key_schedule(std::span<const uint8_t> key) {
+   clear();
+   m_key = as_array_of_uint64<2>(key);
+}
+
+void Ascon_AEAD128_Mode::set_associated_data_n(size_t idx, std::span<const uint8_t> ad) {
+   BOTAN_ARG_CHECK(idx == 0, "Ascon-AEAD128: cannot handle non-zero index in set_associated_data_n");
+
+   BOTAN_STATE_CHECK(has_keying_material());
```

Comment:
This (and `associated_data_requires_key` returning true) doesn't make sense to me - we are just assigning to `m_ad` here without needing keying material afaict

Reply:
I'm guessing this is a left-over from a previous iteration where I wanted to process the AD right away without storing it in the internal member. This would have required the key, of course.